File: src/main/java/hoten/serving/ConnectionToServerHandler.java
package hoten.serving;

import hoten.serving.fileutils.FileUtils;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import java.io.File;
import java.io.IOException;
import java.lang.reflect.Type;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.logging.Level;
import java.util.logging.Logger;

public abstract class ConnectionToServerHandler extends SocketHandler {

    public final File localDataFolder;

    public ConnectionToServerHandler(Socket socket) throws IOException {
        super(socket);
        localDataFolder = new File(_in.readUTF()); // :(
        localDataFolder.mkdirs();
    }

    public void start() {
        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                respondToHashes();
                readNewFilesFromServer();
                onConnectionSettled();
                processDataUntilClosed();
            } catch (IOException ex) {
                Logger.getLogger(ConnectionToServerHandler.class.getName()).log(Level.SEVERE, null, ex);
            }
        });
    }

    private void respondToHashes() throws IOException {
        Map<String, byte[]> hashes = readFileHashesFromServer();
        List<File> localFiles = FileUtils.getAllFilesInDirectory(localDataFolder);
        List<String> filesToRequest = compareFileHashes(localFiles, hashes);
        _out.writeUTF(new Gson().toJson(filesToRequest, List.class));
        localFiles.stream().forEach((f) -> {
            f.delete();
        });
    }

    private Map<String, byte[]> readFileHashesFromServer() throws IOException {
        String jsonHashes = _in.readUTF();
        Type type = new TypeToken<Map<String, byte[]>>() {
        }.getType();
        return new Gson().fromJson(jsonHashes, type);
    }

    private List<String> compareFileHashes(List<File> files, Map<String, byte[]> hashes) {
        List<String> filesToRequest = new ArrayList();
        hashes.forEach((fileName, fileHash) -> {
            File f = new File(localDataFolder, fileName);

            for (File cf : files) {
                if (cf.equals(f)) {
                    //remove this file as a candidate for pruning
                    files.remove(cf);
                    break;
                }
            }

            if (!f.exists() || !Arrays.equals(FileUtils.md5HashFile(f), fileHash)) {
                filesToRequest.add(fileName);
            }
        });
        return filesToRequest;
    }

    private void readNewFilesFromServer() throws IOException {
        int numFiles = _in.readInt();
        for (int i = 0; i < numFiles; i++) {
            String fileName = _in.readUTF();
            int length = _in.readInt();
            Logger.getLogger(ConnectionToServerHandler.class.getName()).log(Level.INFO, "Updating {0}, size = {1}", new Object[]{fileName, length});
            byte[] data = new byte[length];
            _in.readFully(data);
            FileUtils.saveAs(new File(localDataFolder, fileName), data);
        }
        _out.write(0);//done updating files
    }
}

File: src/main/csharp/Serving/Serving/ConnectionToServerHandler.cs
using MiscUtil.Conversion;
using MiscUtil.IO;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Threading;

namespace Serving
{
    public abstract class ConnectionToServerHandler
    {
        public String LocalDataFolder { get; private set; }

        private Thread _dataHandleThread;
        private Socket _socket;
        private JavaBinaryReader _in;
        private JavaBinaryWriter _out;
        private Protocols _protocols;
        private BoundDest _boundTo;
        private BoundDest _boundFrom;

        public ConnectionToServerHandler(String host, int port, Protocols protocols, BoundDest boundTo)
        {
            Connect(host, port);
            _protocols = protocols;
            _boundTo = boundTo;
            _boundFrom = boundTo == BoundDest.CLIENT ? BoundDest.SERVER : BoundDest.CLIENT;
        }

        public void Start()
        {
            _dataHandleThread = new Thread(() =>
            {
                Thread.CurrentThread.IsBackground = true;

                /*String test = "a";
                do { 
                    test = test + test; 
                } while (test.Length < 2000);

                _out.WriteJavaUTF(test);

                _out.Write(123456);
                _out.Write((short)1234);
                _out.Write((byte)123);

                Console.WriteLine("gettting data");
                Console.WriteLine(_in.ReadJavaUTF());
                Console.WriteLine(_in.ReadInt32());
                Console.WriteLine(_in.ReadInt16());
                Console.WriteLine(_in.ReadByte());*/

                LocalDataFolder = _in.ReadJavaUTF(); // :(
                Directory.CreateDirectory(LocalDataFolder);
                RespondToHashes();
                ReadNewFilesFromServer();
                OnConnectionSettled();
                while (true)
                {
                    HandleData();
                }
            });
            _dataHandleThread.Start();
        }

        protected abstract void OnConnectionSettled();

        protected abstract void HandleData(int type, JObject data);

        protected abstract void HandleData(int type, JavaBinaryReader data);

        public void Send(Message message)
        {
            try
            {
                lock (_out)
                {
                    _out.Write(message.Data.Length);
                    _out.Write(message.Protocol.Type);
                    _out.Write(message.Data);
                }
            }
            catch (IOException ex)
            {
                Close();
            }
        }

        public void Close()
        {
            _dataHandleThread.Abort();
            try
            {
                _out.Close();
                _in.Close();
                _socket.Close();
            }
            catch (IOException ex)
            {
                Console.WriteLine("Error closing streams " + ex);
            }
        }

        private void Connect(String host, int port)
        {
            _socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            _socket.Connect(host, port);
            var stream = new NetworkStream(_socket);
            
            _in = new JavaBinaryReader(stream);
            _out = new JavaBinaryWriter(stream);
        }

        private void RespondToHashes()
        {
            var hashes = ReadFileHashesFromServer();
            var localFiles = new List<String>(Directory.GetFiles(LocalDataFolder));
            var filesToRequest = CompareFileHashes(localFiles, hashes);
            var json = JsonConvert.SerializeObject(filesToRequest);
            _out.WriteJavaUTF(json);
        }

        private Dictionary<String, sbyte[]> ReadFileHashesFromServer()
        {
            var jsonHashes = _in.ReadJavaUTF();
            return JsonConvert.DeserializeObject<Dictionary<String, sbyte[]>>(jsonHashes);
        }

        private List<String> CompareFileHashes(List<String> files, Dictionary<String, sbyte[]> hashes)
        {
            var filesToRequest = new List<String>();
            var md5 = MD5.Create();

            foreach (var entry in hashes)
            {
                var fileName = entry.Key;
                var fileHash = entry.Value;
                var path = System.IO.Path.Combine(LocalDataFolder, fileName);

                files.Remove(fileName);
                if (!File.Exists(path))
                {
                    filesToRequest.Add(fileName);
                }
                else
                {
                    using (var stream = File.OpenRead(path))
                    {
                        var localHash = (sbyte[])(Array)md5.ComputeHash(stream);
                        if (!fileHash.SequenceEqual(localHash))
                        {
                            filesToRequest.Add(fileName);
                        }
                    }
                }
            }
            return filesToRequest;
        }

        private void ReadNewFilesFromServer()
        {
            int numFiles = _in.ReadInt32();
            for (int i = 0; i < numFiles; i++)
            {
                var fileName = _in.ReadJavaUTF();
                int length = _in.ReadInt32();
                var data = _in.ReadBytes(length);
                var path = System.IO.Path.Combine(LocalDataFolder, fileName);
                Directory.CreateDirectory(Path.GetDirectoryName(path));
                File.WriteAllBytes(path, data);
            }
            _out.Write((byte)0); //done updating files
        }

        private void HandleData()
        {
            int dataSize = _in.ReadInt32();
            int type = _in.ReadInt32();
            byte[] bytes = _in.ReadBytes(dataSize);
            Message message = Message.InboundMessage(_protocols.Get(_boundFrom, type), bytes);
            Object interpreted = message.Interpret();
            if (interpreted is JObject)
            {
                HandleData(type, interpreted as JObject);
            }
            else if (interpreted is JavaBinaryReader)
            {
                HandleData(type, interpreted as JavaBinaryReader);
            }
        }

        protected Protocol Outbound(Enum protocolEnum)
        {
            return _protocols.Get(_boundTo, Convert.ToInt32(protocolEnum));
        }
    }
}
